src/merkle/tiered_smt/mod.rs

```diff
@@ -44,60 +44,17 @@ mod tests;
 /// - Leaf node at depth 64: hash([key_0, value_0, ..., key_n, value_n], domain=64).
 #[derive(Debug, Clone, PartialEq, Eq)]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
-pub struct TieredSmt {
+pub struct TieredSmt<T> {
     root: RpoDigest,
     nodes: NodeStore,
-    values: ValueStore,
+    values: ValueStore<T>,
 }
 
-impl TieredSmt {
-    // CONSTANTS
-    // --------------------------------------------------------------------------------------------
-
-    /// The number of levels between tiers.
-    const TIER_SIZE: u8 = 16;
-
-    /// Depths at which leaves can exist in a tiered SMT.
-    const TIER_DEPTHS: [u8; 4] = [16, 32, 48, 64];
-
-    /// Maximum node depth. This is also the bottom tier of the tree.
-    const MAX_DEPTH: u8 = 64;
-
-    /// Value of an empty leaf.
-    pub const EMPTY_VALUE: Word = super::empty_roots::EMPTY_WORD;
-
-    // CONSTRUCTORS
-    // --------------------------------------------------------------------------------------------
-
-    /// Returns a new [TieredSmt] instantiated with the specified key-value pairs.
-    ///
-    /// # Errors
-    /// Returns an error if the provided entries contain multiple values for the same key.
-    pub fn with_leaves<R, I>(entries: R) -> Result<Self, MerkleError>
-    where
-        R: IntoIterator<IntoIter = I>,
-        I: Iterator<Item = (RpoDigest, Word)> + ExactSizeIterator,
-    {
-        // create an empty tree
-        let mut tree = Self::default();
-
-        // append leaves to the tree returning an error if a duplicate entry for the same key
-        // is found
-        let mut empty_entries = BTreeSet::new();
-        for (key, value) in entries {
-            let old_value = tree.insert(key, value);
-            if old_value != Self::EMPTY_VALUE || empty_entries.contains(&key) {
-                return Err(MerkleError::DuplicateValuesForKey(key));
-            }
-            // if we've processed an empty entry, add the key to the set of empty entry keys, and
-            // if this key was already in the set, return an error
-            if value == Self::EMPTY_VALUE && !empty_entries.insert(key) {
-                return Err(MerkleError::DuplicateValuesForKey(key));
-            }
-        }
-        Ok(tree)
-    }
+const TIER_DEPTHS: [u8; 4] = [16, 32, 48, 64];
+const TIER_SIZE: u8 = 16;
+const MAX_DEPTH: u8 = 64;
```

Comment:
Had to move the constants out of the `impl` to refer to them inside generic functions, namely `hash_upper_leaf`